In [1]:
pwd()

'C:\\Users\\aparashar'

In [16]:
import codecs
import glob
import regex as re
import pandas as pd
# for converting wx to utf notation
import wxconv
import numpy as np
from wxconv import WXC
import swifter

In [3]:
%cd E:\\META\\ltrc

E:\META\ltrc


In [4]:
def getLines(file,splitby):
    with codecs.open(file, 'r', encoding="utf-8") as filer:
                    data = filer.read()
                    data = re.sub('\t+', '\t', data)
    lines = data.split(splitby)
    return lines

In [5]:
def get_sentence(row):
    ret=re.findall('"(.+?)"', row['gloss'])
    return ret
def get_gloss(row):
    ret = []
    split = row['gloss'].split(';')
    for i in split:
        if '"' not in i:
            ret.append(i)
        else:
            break
    return ret
def getindexdefault(arr,elem, default):
        try:
            thing_index = arr.index(elem)
            return thing_index
        except ValueError:
            return default

### Converting hin-eng linked data to a dataframe


In [6]:
with codecs.open('English_Hindi_Linked_Synsets', 'r', encoding="utf-8") as filer:
                    data = filer.read()
                    data = re.sub('\t+', '\t', data)
with codecs.open('English_Hindi_Linked_Synsets_m.txt', 'w', encoding="utf-8") as filew:
    filew.write(data)
linked_df = pd.read_csv('English_Hindi_Linked_Synsets_m.txt', sep='\t',na_filter = False)
linked_df['synset_words'] = linked_df['synset_words'].str.split(', ')
linked_df['hindi_synset_words'] = ','+linked_df['hindi_synset_words']+','
linked_df['eng_ex'] = linked_df.apply(lambda row: get_sentence(row), axis=1)
linked_df['gloss'] = linked_df.apply(lambda row: get_gloss(row), axis=1)
mask_no_engex = linked_df.eng_ex.apply(lambda x: len(x)==0)
print("Total no. of instances : " + str(len(linked_df)))
print("No. of instances having no english examples: " + str(len(linked_df[mask_no_engex])))
mask = linked_df.hindi_synset_words.apply(lambda x: 'लाना' in x)
linked_df[mask]
#linked_df[mask].to_csv('anka_2.csv', sep ='\t')
#print(len(linked_df[mask]))
linked_df[0:20]

Total no. of instances : 17787
No. of instances having no english examples: 10315


SNo  synset_id   category  \
0     1    1733661  ADJECTIVE   
1     2     935500  ADJECTIVE   
2     3     537682       NOUN   
3     4    1210281       NOUN   
4     5     585660       NOUN   
5     6    1866030       VERB   
6     7   12339831       NOUN   
7     8    7447022       NOUN   
8     9   11448343       NOUN   
9    10   11519121       NOUN   
10   11   10069427       NOUN   
11   12    1020936       NOUN   
12   13    7090019       NOUN   
13   14   10085970       NOUN   
14   15    4058096       NOUN   
15   16    2119971  ADJECTIVE   
16   17     150505  ADJECTIVE   
17   18   10420031       NOUN   
18   19    1380267  ADJECTIVE   
19   20    1380267  ADJECTIVE   

                                         synset_words  \
0                                            [unborn]   
1                            [experienced, experient]   
2                          [folk_dancing, folk_dance]   
3                 [community_service, public_service]   
4                                    [public_service]   
5                                    [wallow, welter]   
6                                             [clove]   
7                               [return, coming_back]   
8                       [east_wind, easter, easterly]   
9              [thunderbolt, bolt, bolt_of_lightning]   
10                  [executioner, public_executioner]   
11                                        [imitation]   
12                              [verbiage, verbalism]   
13                             [fiancee, bride-to-be]   
14                                     [reading_room]   
15                       [thoughtful, serious-minded]   
16  [eloquent, facile, fluent, silver, silver-tong...   
17     [petitioner, suppliant, supplicant, requester]   
18                                           [aerial]   
19                                           [aerial]   

                                   hindi_synset_words hindi_synset_id  \
0   ,अजन्मा,अजात,अनुत्पन्न,अनुद्भूत,अप्रादुर्भूत,अ...               1   
1   ,अनुभवी,जानकार,परिपक्व,तजुर्बेकार,तजुरबेकार,तज...             100   
2                      ,लोकनृत्य,लोक-नृत्य,लोक नृत्य,           10001   
3                                    ,लोकसेवा,जनसेवा,           10002   
4                                    ,लोकसेवा,जनसेवा,           10003   
5                                             ,लोटना,           10006   
6   ,लौंग,लवंग,लोंग,श्रीसंज्ञ,रुचिर,मादन,अमरकुसुम,...           10008   
7                                              ,null,           10009   
8   ,पुरवाई,पुरवा,पुरुआ,पुरवैया,पुरबिया हवा,पूर्व ...            1001   
9          ,वज्रपात,वज्राघात,विद्युत्पात,विद्युत पात,           10012   
10                                  ,जल्लाद,वधिक,वधक,           10016   
11  ,अनुकरण,अनुसरण,अनुगमन,नक़ल,नकल,अनुकार,अनुसार,अ...            1002   
12             ,शब्दजाल,शब्दाडंबर,शब्दाडम्बर,वाक्जाल,           10022   
13                                        ,वाग्दत्ता,           10024   
14                                          ,वाचनालय,           10025   
15  ,विचारशील,विचारवान,दीर्घदर्शी,विचारवान्,बुद्धि...           10026   
16                                          ,वाक्पटु,           10030   
17  ,वादी,फरियादी,मुद्दई,अभियोक्ता,अभियोगकर्ता,अभि...           10032   
18                                           ,वायव्य,           10033   
19                                           ,वायव्य,           10033   

   hindi_category                                              gloss  \
0       adjective                   [not yet brought into existence]   
1       adjective  [having experience,  having knowledge or skill...   
2            noun  [a style of dancing that originated among ordi...   
3            noun  [a service that is performed for the benefit o...   
4            noun  [employment within a government system (especi...   
5            verb                                                 []   
6            noun  [aromatic flower bud of a clove tree,  yields ...   
7 

### Converting hindi synset data to a dataframe

In [7]:
d = {'ID':[],'CAT':[],'CONCEPT':[],'EXAMPLE':[],'SYNSET-HINDI':[]}
for line in getLines('hindi.syns','\n\n'):
    internal_lines = line.split('\n')
    #print(internal_lines)
    if len(internal_lines)>0:
        
        d['ID'].append(internal_lines[0].split('::')[1].strip())
        d['CAT'].append(internal_lines[1].split('::')[1].strip())
        d['CONCEPT'].append(internal_lines[2].split('::')[1].strip())
        d['EXAMPLE'].append(internal_lines[3].split('::')[1].strip())
        d['SYNSET-HINDI'].append(':'+internal_lines[4].split('::')[1].strip()+':')

hn_syns_df = pd.DataFrame.from_dict(d)
print(len(hn_syns_df))
mask3 = hn_syns_df['SYNSET-HINDI'].apply(lambda x: ':अपवित्र_स्थान:' in x)
cross = hn_syns_df[mask3]
cross
#hn_syns_df[:10]

38759


ID   CAT                  CONCEPT  \
5  6  NOUN  वह स्थान जो पवित्र न हो   

                                             EXAMPLE  \
5  "धार्मिक मान्यता है कि भूत-प्रेत अपवित्र स्थान...   

                              SYNSET-HINDI  
5  :अपवित्र_स्थान:अपुण्यभूमि:अपवित्रस्थली:

### Converting concept-dict(mrs) into a dataframe

In [8]:
d ={'conc_label_wx':[],'conc_label_utf':[],'conc_in_eng':[],'mrs_conc':[]}
con = WXC(order='wx2utf', lang='hin')
for line in getLines('H_concept-to-mrs-rels.dat','\n'):
    internal_lines = line.split('\t')
    if len(internal_lines)>=4:
        d['conc_label_wx'].append(internal_lines[1])
        d['conc_label_utf'].append(con.convert(internal_lines[1]).replace('+',' '))
        d['conc_in_eng'].append(internal_lines[2])
        d['mrs_conc'].append(internal_lines[3])
    else:
        print(internal_lines[2])
        d['conc_label_wx'].append(internal_lines[1])
        d['conc_label_utf'].append(con.convert(internal_lines[1]).replace('+',' '))
        if len(internal_lines[2])>20:
            d['conc_in_eng'].append(internal_lines[2].split(' ')[0])
            d['mrs_conc'].append(internal_lines[2].split(' ')[1])
        else:
            d['conc_in_eng'].append(internal_lines[2][0:7])
            d['mrs_conc'].append(internal_lines[2][8:])
mrs_df = pd.DataFrame.from_dict(d)
print( "No. of instances in mrs file : "+ str(len(mrs_df)))
#mask1 = mrs_df.conc_label_utf.apply(lambda x: 'हो_' in x)
#mask2 = mrs_df.mrs_conc.apply(lambda x: '_v_' in x)
#mrs_df[mask1 & mask2]
mrs_df[:10]

foundation+stone_1 _foundation_n_1+_stone_n_1)
dutch_1_dutch_a_1_1)
shrug_3_shrug_n_1_1)
No. of instances in mrs file : 54329


conc_label_wx conc_label_utf   conc_in_eng                  mrs_conc
0           A_1            आ_1        come_1                _come_v_1)
1         Aba_1           आब_1      lustre_1     _lustre/NN_u_unknown)
2         ABA_1          आभा_1      lustre_2     _lustre/NN_u_unknown)
3         aba_1           अब_1         now_3           def_implicit_q)
4         aba_1           अब_1         now_3                loc_nonsp)
5         aba_1           अब_1         now_3                 _now_a_1)
6         aba_1           अब_1         now_3                   time_n)
7         Aba_2           आब_2  brilliance_1          _brilliance_n_1)
8         ABA_2          आभा_2   splendour_1  _splendour/NN_u_unknown)
9         ABA_3          आभा_3        tint_1               _tint_n_of)

###  Generating the final linked data set 
#### [hin_word, eng_conc]  from the concept-dict are used to filter the hin_eng linked dataset 

In [17]:
con2 = WXC(order='utf2wx')
def get_fin_df(row,d,ign,c):
    conc_label_hn = row['conc_label_utf']
    conc_eng = row['conc_in_eng']
    mrs_conc = row['mrs_conc']
    hn_label =','+ conc_label_hn.split('_')[0]+','
    eng_label = conc_eng.split('_')[0]
    # checking if the hindi label is present in the hindi synset words list of linked_df
    #if the word is a verb
    if '_v_' in row['mrs_conc']:
        # adding 'ना' suffix before filtering
        hn_label_na = ','+conc_label_hn.split('_')[0] + 'ना' +','
        # adding 'करना' suffix before filtering
        hn_label_krna = ','+conc_label_hn.split('_')[0] + ' करना' + ','
        # adding 'होना' suffix before filtering
        hn_label_hona = ','+conc_label_hn.split('_')[0] + ' होना' + ','
        mask = linked_df.hindi_synset_words.swifter.progress_bar(False).apply(lambda x: (hn_label in x) or (hn_label_na in x) or (hn_label_krna in x) or (hn_label_hona in x))
    # if the word is not a verb
    else:
        mask = linked_df.hindi_synset_words.swifter.progress_bar(False).apply(lambda x: hn_label in x)

    
    #mask = linked_df.hindi_synset_words.apply(lambda x: (hn_label in x) or (hn_label_na in x) or (hn_label_krna in x))
    # checking if the eng label is present in the eng sysnet word list of linked_df
    mask1 = linked_df.synset_words.swifter.progress_bar(False).apply(lambda x: eng_label in x)
    # taking intersection of both the conditions
    cross = linked_df[mask & mask1]
    cross = cross.reset_index()  
    # if this pair of [hindi_label and conc_eng] are found in the same row of linked_df
    if cross.empty == False:
        
        for index, row in cross.iterrows():
            #if (getindexdefault(d['eng_id'],row['synset_id'],-1)!=getindexdefault(d['hin_id'],row['hindi_synset_id'],-1)):
            #if (row['synset_id'] not in d['eng_id']) and (row['hindi_synset_id'] not in d['hin_id']):
                # setting a flag signifying that the hindi word and conc eng pair were found in the same row
                #d['ign'].append(False)
                
                d['hin'].append(conc_label_hn)
                d['eng'].append(conc_eng)
                d['eng_id'].append(row['synset_id'])
                d['hin_id'].append(row['hindi_synset_id'])
                d['eng_gloss'].append(row['gloss'])
                d['eng_ex'].append(row['eng_ex'])
                if len(hn_syns_df[hn_syns_df['ID']==row['hindi_synset_id']]) != 0:
                    # getting hindi example and concept from the hindi synset dataframe
                    d['hin_ex'].append(hn_syns_df[hn_syns_df['ID']==row['hindi_synset_id']].EXAMPLE.item())
                    d['hin_gloss'].append(hn_syns_df[hn_syns_df['ID']==row['hindi_synset_id']].CONCEPT.item())
                else:
                    # if the hindi synset id is not found in hindi synset df. then adding empty strings
                    d['hin_ex'].append("")
                    d['hin_gloss'].append("")
                hn_syn_list = row['hindi_synset_words'].split(',')
                for word in hn_syn_list:
                    c['conc_label_utf'].append(word)
                    c['conc_label_wx'].append((con2.convert(word)))
                    c['conc_in_eng'].append(conc_eng)
                    c['mrs_conc'].append(mrs_conc)
    
    # else taking the corresponding hindi and english details(gloss,id, examples) separately in a many to many fashion
    else:
        
        # adding hindi-word and eng_concept pairs not found in the intersection to the ign dict
        ign['hindi_ig'].append(conc_label_hn)
        ign['eng_conc_ig'].append(conc_eng)
        ign['mrs_conc'].append(row['mrs_conc'])
        
                

In [18]:
d = {'hin' : [],'eng' : [], 'hin_id':[],'eng_id':[], 'hin_gloss':[],'eng_gloss':[], 'hin_ex':[],'eng_ex':[]}
c = {'conc_label_wx':[],'conc_label_utf':[],'conc_in_eng':[],'mrs_conc':[]}
ign = {'hindi_ig' :[], 'eng_conc_ig':[],'mrs_conc':[]}
mrs_df.apply(lambda row: get_fin_df(row,d,ign,c), axis=1)
# final dataframe depending on the hindi label and eng conc from mrs
fin_df = pd.DataFrame.from_dict(d)
#ignored pairs which were not found in eng-hindi linked df
ign_df = pd.DataFrame.from_dict(ign)


In [67]:
def count_mrs(row,h):
    if row['conc_label_wx'] in h.keys():
        h[row['conc_label_wx']] +=1
        row['conc_label_wx'] += '_' + str(h[row['conc_label_wx']])
        row['conc_label_wx'] = row['conc_label_wx'].replace(' ','+')
    else:
        h[row['conc_label_wx']] = 1
        row['conc_label_wx'] += '_' + str(h[row['conc_label_wx']])
        row['conc_label_wx'] = row['conc_label_wx'].replace(' ','+')

    return row['conc_label_wx']

In [68]:
aug_mrs_df = pd.DataFrame.from_dict(c)
aug_mrs_df = aug_mrs_df[aug_mrs_df['conc_label_wx']!='']
aug_mrs_df = aug_mrs_df.drop_duplicates(subset=['conc_label_wx','conc_label_utf','conc_in_eng','mrs_conc'], keep='first').reset_index()
mr = aug_mrs_df
mr

index  conc_label_wx conc_label_utf conc_in_eng         mrs_conc
0            1            AnA            आना      come_1       _come_v_1)
1            7            aBI            अभी       now_3  def_implicit_q)
2            8      PZilahAla        फ़िलहाल       now_3  def_implicit_q)
3            9       PilahAla         फिलहाल       now_3  def_implicit_q)
4           10     isa samaya         इस समय       now_3  def_implicit_q)
...        ...            ...            ...         ...              ...
139406  178087         aba BI          अब भी    still_16      _still_a_1)
139407  178090    jArI raKanA      जारी रखना  continue_2   _continue_v_2)
139408  178091    SurU raKanA      शुरू रखना  continue_2   _continue_v_2)
139409  178092    cAlU raKanA      चालू रखना  continue_2   _continue_v_2)
139410  178093  kAyama raKanA      कायम रखना  continue_2   _continue_v_2)

[139411 rows x 5 columns]

In [69]:
h = dict()
mr['conc_label_wx'] = mr.apply(lambda row: count_mrs(row,h), axis=1)
mr['conc_label_utf'] = mr.apply(lambda row: con.convert(row['conc_label_wx']), axis=1) 
mr

index    conc_label_wx conc_label_utf conc_in_eng         mrs_conc
0            1            AnA_1          आना_1      come_1       _come_v_1)
1            7            aBI_1          अभी_1       now_3  def_implicit_q)
2            8      PZilahAla_1      फ़िलहाल_1       now_3  def_implicit_q)
3            9       PilahAla_1       फिलहाल_1       now_3  def_implicit_q)
4           10     isa+samaya_1       इस+समय_1       now_3  def_implicit_q)
...        ...              ...            ...         ...              ...
139406  178087         aba+BI_1        अब+भी_1    still_16      _still_a_1)
139407  178090    jArI+raKanA_1    जारी+रखना_1  continue_2   _continue_v_2)
139408  178091    SurU+raKanA_1    शुरू+रखना_1  continue_2   _continue_v_2)
139409  178092    cAlU+raKanA_1    चालू+रखना_1  continue_2   _continue_v_2)
139410  178093  kAyama+raKanA_1    कायम+रखना_1  continue_2   _continue_v_2)

[139411 rows x 5 columns]

In [72]:
mr.reset_index(drop=True)
mr.drop(columns=['index'])

conc_label_wx conc_label_utf conc_in_eng         mrs_conc
0                 AnA_1          आना_1      come_1       _come_v_1)
1                 aBI_1          अभी_1       now_3  def_implicit_q)
2           PZilahAla_1      फ़िलहाल_1       now_3  def_implicit_q)
3            PilahAla_1       फिलहाल_1       now_3  def_implicit_q)
4          isa+samaya_1       इस+समय_1       now_3  def_implicit_q)
...                 ...            ...         ...              ...
139406         aba+BI_1        अब+भी_1    still_16      _still_a_1)
139407    jArI+raKanA_1    जारी+रखना_1  continue_2   _continue_v_2)
139408    SurU+raKanA_1    शुरू+रखना_1  continue_2   _continue_v_2)
139409    cAlU+raKanA_1    चालू+रखना_1  continue_2   _continue_v_2)
139410  kAyama+raKanA_1    कायम+रखना_1  continue_2   _continue_v_2)

[139411 rows x 4 columns]

In [78]:
mr.to_csv('augmented_mrs_v1.csv', sep ='\t')


In [205]:
mask_no_engex = fin_df.eng_ex.apply(lambda x: len(x)==0)
mask_no_hin_id = fin_df.hin_ex.apply(lambda x: x=="")
# dropping duplicate rows having same word,concept and synset id's
fin_df = fin_df.drop_duplicates(subset=['hin','eng','hin_id','eng_id'], keep='first').reset_index()
print("No. of total instances: " + str(len(fin_df)))
print("No. of instances having no corresponding hindi ids :" + str(len(fin_df[mask_no_hin_id])))
print("No. of instances having no english examples :" + str(len(fin_df[mask_no_engex])))
print("No. of ignored pairs from the MRS: " + str(len(ign_df)))
fin_df[:20]

No. of total instances: 14209
No. of instances having no corresponding hindi ids :169
No. of instances having no english examples :6689
No. of ignored pairs from the MRS: 41850


level_0  index        hin            eng hin_id    eng_id  \
0         0      0        आ_1         come_1  30934   2729963   
1         1      1        आ_1         come_1  11882    341917   
2         2      2       अब_1          now_3  11083     49220   
3         3      3       अब_1          now_3  23773     49102   
4         4      4       अब_1          now_3   2785     48475   
5         5     14       आब_3        water_2   1930  14845743   
6         6     15   अबाबील_1      swallow_1   4233   1594372   
7         7     16    अभागा_1  unfortunate_1   2279   1049462   
8         8     17    अभागा_1  unfortunate_1  23230   9630641   
9         9     18   अभाग्य_1   misfortune_3   2676  14475405   
10       10     19    आबहवा_1      climate_1   5520  14519366   
11       11     20  अभक्ष्य_1    uneatable_2  14745    829496   
12       12     21  अभक्ष्य_1    uneatable_2   2334    829496   
13       13     22  अभक्ष्य_2     inedible_2  14745    829496   
14       14     23  अभक्ष्य_2     inedible_2   2334    829496   
15       15     24    आबंटन_1    allotment_3  13650   1083645   
16       16     25    आबनूस_1        ebony_1  26604  12770529   
17       17     26   आबपाशी_1   irrigation_1   1869   1057631   
18       18     27     आभार_2    gratitude_1   3459   7504343   
19       19     28     आभार_4       burden_1   3460   5832745   

                                            hin_gloss  \
0                                किसी के अंतर्गत होना   
1              किसी भाव या अवस्था आदि का उत्पन्न होना   
2                      इसी समय या इस वक़्त या इस क्षण   
3                इस समय (विषय या कार्य परिवर्तन सूचक)   
4                           आधुनिक या वर्तमान समय में   
5   नदी, जलाशय, वर्षा आदि से मिलने वाला वह द्रव पद...   
6           एक प्रकार का पक्षी जो काले रंग का होता है   
7                                   जो भाग्यशाली न हो   
8                        वह व्यक्ति जो भाग्यशाली न हो   
9                                   मंद या बुरा भाग्य   
10  किसी स्थान की वह प्राकृतिक स्थिति जिसका प्राणि...   
11                               जिसे खाना नहीं चाहिए   
12                              जो खाने के योग्य न हो   
13                               जिसे खाना नहीं चाहिए   
14                              जो खाने के योग्य न हो   
15                                   बाँटने की क्रिया   
16  एक पेड़ जो विशेषकर दक्षिण एशिया में पाया जाता ...   
17  खेती-बारी के लिए खेतों आदि में नाली आदि के द्व...   
18                       कृतज्ञ होने की अवस्था या भाव   
19  किसी की जिम्मेदारी बनकर रहने तथा उसके लिए कुछ ...   

                                            eng_gloss  \
0             [come under, be classified or included]   
1           [come to pass,  arrive, as in due course]   
2                             [at the present moment]   
3   [used to preface a command or reproof or request]   
4                                    [in these times]   
5   [binary compound that occurs at room temperatu...   
6   [small long-winged songbird noted for swift gr...   
7   [not favored by fortune,  marked or accompanie...   
8                   [a person who suffers misfortune]   
9   [an unfortunate state resulting from unfavorab...   
10  [the weather in some location averaged over so...   
11                            [not suitable for food]   
12                            [not suitable for food]   
13                            [not suitable for food]   
14                            [not suitable for food]   
15  [the act of distributing by allotting or appor...   
16  [tropical tree of southern Asia having hard da...   
17  [supplying dry land with water by means of dit...   
18       [a feeling of thankfulness and appreciation]   
19                  [an onerous or difficult concern]   

                                               hin_ex  \
0   "बनारस उत्तर-प्रदेश में आता है । / यह कथा रामा...   
1       "आज के हास्य कवि सम्मेलन में बहुत आनंद आया ।"   
2                          "अभी मैं सोना चाहता हूँ ।"   
3             

###  Concept-dict (mrs) hindi words and concepts not found in the linked-synset data

In [74]:
# checking which hindi and english words from mrs are not present in the synsets at all
def notFound(row,hin_nf):
    conc_label_hn = row['conc_label_utf']
    conc_eng = row['conc_in_eng']
    hn_label = ','+conc_label_hn.split('_')[0]+','
    hn_label = hn_label.replace('+',' ')
    
    hn_label_hs = ':' + conc_label_hn.split('_')[0]+':'
    hn_label_hs = hn_label_hs.replace('+','_')
    eng_label = conc_eng.split('_')[0]
    # checking ig the hindi label is present in the hindi synset words list of linked_df
    #mask = linked_df.hindi_synset_words.apply(lambda x: hn_label in x)
    if '_v_' in row['mrs_conc']:
        # adding 'ना' suffix before filtering
        hn_label_na = ','+conc_label_hn.split('_')[0] + 'ना' +','
        hn_label_hs_na = ':'+conc_label_hn.split('_')[0] + 'ना' +':'
        # adding 'करना' suffix before filtering
        hn_label_krna = ','+conc_label_hn.split('_')[0] + ' करना' + ','
        hn_label_hs_krna = ':'+conc_label_hn.split('_')[0] + ' करना' + ':'
        # adding 'होना' suffix before filtering
        hn_label_hona = ','+conc_label_hn.split('_')[0] + ' होना' + ','
        hn_label_hs_hona = ':'+conc_label_hn.split('_')[0] + ' होना' + ':'
        mask = linked_df.hindi_synset_words.apply(lambda x: (hn_label in x) or (hn_label_na in x) or (hn_label_krna in x) or (hn_label_hona in x))
        mask3 = hn_syns_df['SYNSET-HINDI'].apply(lambda x: (hn_label_hs in x) or (hn_label_hs_na in x) or (hn_label_hs_krna in x) or (hn_label_hs_hona in x))
    # if the word is not a verb
    else:
        mask = linked_df.hindi_synset_words.apply(lambda x: hn_label in x)
        mask3 = hn_syns_df['SYNSET-HINDI'].apply(lambda x: hn_label_hs in x)

    # checking if the eng label is present in the eng sysnet word list of linked_df
    #mask1 = linked_df.synset_words.apply(lambda x: eng_label in x)
    hin_masked = linked_df[mask]
    #eng_masked = linked_df[mask1]
    if hin_masked.empty == True:
        
        #mask3 = hn_syns_df['SYNSET-HINDI'].apply(lambda x: hn_label_hs in x)
        cross = hn_syns_df[mask3]
        if cross.empty == False:
            for index,row1 in cross.iterrows():
                hin_nf['hn_wx'].append(row['conc_label_wx'])
                hin_nf['hn_utf'].append(conc_label_hn)
                hin_nf['conc_eng'].append(conc_eng)
                hin_nf['conc_hn'].append(row1['CONCEPT'])
                hin_nf['hn_ex'].append(row1['EXAMPLE'])
    #if eng_masked.empty == True:
      #  eng_nf.append(eng_label)
    

In [75]:
hin_nf = {'hn_wx' :[],'hn_utf':[],'conc_eng':[],'conc_hn':[],'hn_ex':[]}
eng_nf = {}
mrs_df.apply(lambda row: notFound(row,hin_nf), axis=1)
#hin_nf

0        None
1        None
2        None
3        None
4        None
         ... 
54324    None
54325    None
54326    None
54327    None
54328    None
Length: 54329, dtype: object

In [76]:
hindi_nf_df =pd.DataFrame.from_dict(hin_nf)
#hindi_nf_df = hindi_nf_df.drop_duplicates(subset=['hindi_word'], keep='first').reset_index()
hindi_nf_df

hn_wx      hn_utf          conc_eng  \
0     AbakArI_1    आबकारी_1          excise_1   
1     AbakArI_1    आबकारी_1          excise_1   
2     aBAviwa_1    अभावित_1  uncontemplated_1   
3     aBicAra_1    अभिचार_1     incantation_1   
4     aBicAra_2    अभिचार_2         sorcery_1   
...         ...         ...               ...   
3064  yuyuwsu_1  युयुत्सु_1      pugnacious_3   
3065  yuyuwsu_2  युयुत्सु_2       bellicose_1   
3066  yuyuwsu_2  युयुत्सु_2       bellicose_1   
3067  yuyuwsu_3  युयुत्सु_3     belligerent_1   
3068  yuyuwsu_3  युयुत्सु_3     belligerent_1   

                                                conc_hn  \
0                             शराब बनने और बिकने की जगह   
1        मादक वस्तुओं से संबंध रखनेवाला एक सरकारी विभाग   
2                    जिसकी भावना या चिंतन न किया गया हो   
3     मंत्र-तंत्र द्वारा मारन तथा उच्चाटन आदि हिंसक ...   
4     मंत्र-तंत्र द्वारा मारन तथा उच्चाटन आदि हिंसक ...   
...                                                 ...   
3064                             धृतराष्ट्र का एक पुत्र   
3065                            युद्ध की इच्छा करनेवाला   
3066                             धृतराष्ट्र का एक पुत्र   
3067                            युद्ध की इच्छा करनेवाला   
3068                             धृतराष्ट्र का एक पुत्र   

                                                  hn_ex  
0                       "रामू शराबघर में काम करता है ।"  
1     "महेश आबकारी में काम करता है । / आज आबकारी मंत...  
2       "वह अभावित विषय को समझने का प्रयास कर रहा था ।"  
3     "अभिचार को प्रमाणित करना सम्भव नहीं हो पाता है ।"  
4     "अभिचार को प्रमाणित करना सम्भव नहीं हो पाता है ।"  
...                                                 ...  
3064         "युयुत्सु का वर्णन महाभारत में मिलता है ।"  
3065  "युद्धेच्छुक राजा ने बिना सोचे-समझे पड़ोसी देश...  
3066         "युयुत्सु का वर्णन महाभारत में मिलता है ।"  
3067  "युद्धेच्छुक राजा ने बिना सोचे-समझे पड़ोसी देश...  
3068         "युयुत्सु का वर्णन महाभारत में मिलता है ।"  

[3069 rows x 5 columns]

In [77]:
hindi_nf_df.to_csv('hindi_nf_set_v1.csv', sep ='\t')


In [209]:
x=np.array(hin_nf)
hin_nf = np.unique(x)
hindi_nf_df = pd.DataFrame({'hindi_word': hin_nf[:]})
hindi_nf_df = hindi_nf_df.drop_duplicates(subset=['hindi_word'], keep='first').reset_index()

y = np.array(eng_nf)
eng_nf = np.unique(y)
eng_nf_df = pd.DataFrame({'eng_word': eng_nf[:]})
eng_nf_df = eng_nf_df.drop_duplicates(subset=['eng_word'], keep='first').reset_index()
# saving hindi not found(hindi_nf) and eng not found(eng_nf) as a csv
eng_nf_df.to_csv('eng_nf.csv', sep ='\t')
hindi_nf_df.to_csv('hin_nf.csv', sep ='\t')
print("Unique hindi words not found in the hin-eng linked synset data:" + str(len(hindi_nf_df)))
print("Unique english words not found in the hin-eng linked synset data:" + str(len(eng_nf_df)))
print(eng_nf_df[:10])
print(hindi_nf_df[:10])

Unique hindi words not found in the hin-eng linked synset data:14369
Unique english words not found in the hin-eng linked synset data:8764
   index  eng_word
0      0          
1      1       -as
2      2  -binding
3      3     -gave
4      4     -inch
5      5      -ist
6      6  -posture
7      7      -tie
8      8   Anatomy
9      9  Anglican
   index    hindi_word
0      0       ,अँकुर,
1      1  ,अँगरेज़ियत,
2      2      ,अँगहीन,
3      3      ,अँगौछा,
4      4   ,अँग्रेज़ी,
5      5       ,अँडुआ,
6      6      ,अँधौटी,
7      7      ,अंक पा,
8      8     ,अंक बना,
9      9    ,अंक लिखा,


### Analyzing [word:meaning] pairs which were not found in the same row in hin-eng linked dataset

In [224]:
print("No. of ignored pairs from the MRS: " + str(len(ign_df)))
ign_df[:10]

No. of ignored pairs from the MRS: 41850


hindi_ig   eng_conc_ig                  mrs_conc
0      आब_1      lustre_1     _lustre/NN_u_unknown)
1     आभा_1      lustre_2     _lustre/NN_u_unknown)
2      आब_2  brilliance_1          _brilliance_n_1)
3     आभा_2   splendour_1  _splendour/NN_u_unknown)
4     आभा_3        tint_1               _tint_n_of)
5     आभा_4       tinge_1               _tinge_n_1)
6   अभागा_2    accursed_1  _accursed/VBN_u_unknown)
7   अभागा_3     unlucky_1             _unlucky_a_1)
8  अभागिन_1    accursed_2  _accursed/VBN_u_unknown)
9  आबकारी_1      excise_1     _excise/NN_u_unknown)

### Checking if any yet unencountered eng synsets exist in the hin-eng linked dataset for ignored hindi words.

In [225]:
def get_coarse_engsyn(row,d):
    conc_label_hn = row['hindi_ig']
    conc_eng = row['eng_conc_ig']
    hn_label =','+ conc_label_hn.split('_')[0]+','
    hn_label = hn_label.replace('+',' ')
    eng_label = conc_eng.split('_')[0]
    # checking ig the hindi label is present in the hindi synset words list of linked_df
    if '_v_' in row['mrs_conc']:
        # adding 'ना' suffix before filtering
        hn_label_na = ','+conc_label_hn.split('_')[0] + 'ना' +','
        # adding 'करना' suffix before filtering
        hn_label_krna = ','+conc_label_hn.split('_')[0] + ' करना' + ','
        # adding 'होना' suffix before filtering
        hn_label_hona = ','+conc_label_hn.split('_')[0] + ' होना' + ','
        mask = linked_df.hindi_synset_words.apply(lambda x: (hn_label in x) or (hn_label_na in x) or (hn_label_krna in x) or (hn_label_hona in x))
    # if the word is not a verb
    else:
        mask = linked_df.hindi_synset_words.apply(lambda x: hn_label in x)
    cross = linked_df[mask]
    if cross.empty == False:
        for index,row in cross.iterrows():
            d['hin'].append(conc_label_hn)
            d['eng_syns'].append(row['synset_words'])
            d['eng_id'].append(row['synset_id'])
            d['hin_id'].append(row['hindi_synset_id'])
            d['eng_gloss'].append(row['gloss'])
            d['eng_ex'].append(row['eng_ex'])
            if len(hn_syns_df[hn_syns_df['ID']==row['hindi_synset_id']]) != 0:
                # getting hindi example and concept from the hindi synset dataframe
                d['hin_ex'].append(hn_syns_df[hn_syns_df['ID']==row['hindi_synset_id']].EXAMPLE.item())
                d['hin_gloss'].append(hn_syns_df[hn_syns_df['ID']==row['hindi_synset_id']].CONCEPT.item())
            else:
                # if the hindi synset id is not found in hindi synset df. then adding empty strings
                d['hin_ex'].append("")
                d['hin_gloss'].append("")
                    
                

In [226]:
d = {'hin' : [],'eng_syns' : [], 'hin_id':[],'eng_id':[], 'hin_gloss':[],'eng_gloss':[], 'hin_ex':[],'eng_ex':[]}
ign_df.swifter.progress_bar(False).apply(lambda row: get_coarse_engsyn(row,d), axis=1)
coarse_syn_df = pd.DataFrame.from_dict(d)

In [ ]:
coarse_syn_df['hin_word'] = coarse_syn_df.swifter.progress_bar(False).apply(lambda row: row['hin'].split('_')[0], axis=1)
coarse_syn_df = coarse_syn_df.drop_duplicates(subset=['hin_word','hin_id','eng_id'], keep='first').reset_index()

#### Removing those rows from coarse_syn_df which have already been encountered in our final_linked_df. 
For example the  {आब_1	[water, H2O]	1930	14845743} has already been added to the original df for a meaning (water) of आब_3 

In [231]:
import swifter
def drop_enc_rows(row):
    conc_label_hn = row['hin']
    hn_label = conc_label_hn.split('_')[0]
    hn_label = hn_label.replace('+',' ')
    #print(hn_label)
    mask1 = fin_df.hin.swifter.progress_bar(False).apply(lambda x: hn_label == x.split('_')[0])
    mask2 = fin_df.hin_id.swifter.progress_bar(False).apply(lambda x: row['hin_id'] == x)
    mask3 = fin_df.eng_id.swifter.progress_bar(False).apply(lambda x : row['eng_id'] == x)
    if fin_df[mask1 & mask2 & mask3].empty == False:
        return True
    else:
        return False   

In [232]:
ten = coarse_syn_df[11:20]
# adding a column for marking if this row already exists in the final_linked_df
coarse_syn_df['exists'] = coarse_syn_df.apply(lambda row:1 if drop_enc_rows(row) else 0, axis=1)
coarse_syn_df

level_0  index             hin  \
0            0      0           अभी_2   
1            1      1        अभिराम_1   
2            2      2       अभ्यासी_3   
3            3      3       अभ्यासी_3   
4            4      4     अचानक गिर_1   
...        ...    ...             ...   
19546    19546  49818  युद्धक्षेत्र_1   
19547    19547  49820     युद्धनीति_1   
19548    19548  49821      युयुत्सा_1   
19549    19549  49823           जनवरी   
19550    19550  49824           फरवरी   

                                                eng_syns hin_id    eng_id  \
0                                      [now, at_present]  11083     49220   
1                                           [attractive]   9106    166146   
2                               [accomplished, complete]     29   2226028   
3      [ace, adept, champion, sensation, maven, mavin...  29929   9762509   
4                                  [spill, tumble, fall]  28493     76884   
...                                                  ...    ...       ...   
19546  [battlefield, battleground, field_of_battle, f...   4741   8506641   
19547                                      [battle_plan]  13549   5902786   
19548                       [bellicosity, bellicoseness]  24391   4643875   
19549                                     [January, Jan]  10846  15210045   
19550                                    [February, Feb]  10851  15210486   

                                               hin_gloss  \
0                         इसी समय या इस वक़्त या इस क्षण   
1                   देखने में अच्छा और मन को लुभाने वाला   
2        जो किसी कार्य को करने में विशेष योग्यता रखता हो   
3      वह जो किसी कार्य को करने की विशेष योग्यता रखता हो   
4               * किसी ऊँची जगह से अचानक गिरने की क्रिया   
...                                                  ...   
19546            वह क्षेत्र जहाँ युद्ध हुआ हो या होता हो   
19547                           युद्ध या रण करने की नीति   
19548                                युद्ध करने की इच्छा   
19549                       अंग्रेज़ी वर्ष का पहला महीना   
19550                        अंग्रेजी साल का दूसरा महीना   

                                               eng_gloss  \
0                                [at the present moment]   
1      [pleasing to the eye or mind especially throug...   
2                                       [highly skilled]   
3       [someone who is dazzlingly skilled in any field]   
4               [a sudden drop from an upright position]   
...                                                  ...   
19546  [a region where a battle is being (or has been...   
19547                     [plan for conducting a battle]   
19548                   [a natural disposition to fight]   
19549  [the first month of the year,  begins 10 days ...   
19550  [the month following January and preceding March]   

                                                  hin_ex  \
0                             "अभी मैं सोना चाहता हूँ ।"   
1      "उसकी चित्रकारी आकर्षक है । / मुझे प्रकृति से ...   
2      "धनुर्विद्या में प्रवीण अर्जुन ने तेल में मछली...   
3             "प्रवीणों को प्रशिक्षण की आवश्यकता नहीं ।"   
4      "पहाड़ी पर से अचानक गिराव के कारण उसे अत्यधिक ...   
...                                                  ...   
19546          "वह अंतिम समय तक युद्धभूमि में डटा रहा ।"   
19547                  "उन्होंने आक्रामक रणनीति अपनाई ।"   
19548  "राजा के मन में युयुत्सा जागृत हुई और उसने पड़...   
19549                     "उसका जन्म जनवरी में हुआ था ।"   
19550           "फरवरी अठाइस या उन्तीस दिन की होती है ।"   

                                                  eng_ex      hin_word  exists  
0      [goods now on sale, the now-aging dictator, th...           अभी       0  
1      [a remarkably attractive young man, an attract...        अभिराम       0  
2         [an accomplished pianist, a complete musician]       अभ्यासी       0  
3                                                     []       अभ्यासी       0  
4                    

In [234]:
# filtering if it does not already exist in the final_linked_df
fin_coarse_syn_df = coarse_syn_df[coarse_syn_df['exists']==0]

In [237]:
fin_coarse_syn_df[fin_coarse_syn_df['hin_word'] =='आब' ]

level_0  index   hin                          eng_syns hin_id   eng_id  \
965      965   3091  आब_1  [sparkle, twinkle, spark, light]   3454  4953380   

            hin_gloss                                          eng_gloss  \
965  एक तरह का प्रकाश  [merriment expressed by a brightness or gleam ...   

                                      hin_ex  \
965  "उसके चेहरे की चमक स्पष्ट झलक रही थी ।"   

                                                eng_ex hin_word  exists  
965  [he had a sparkle in his eye, there' s a perpe...       आब       0

In [ ]:
fin_coarse_syn_df = fin_coarse_syn_df.drop(['level_0','hin_word','exists'],axis=1)

In [251]:
fin_coarse_syn_df = fin_coarse_syn_df.reset_index()

level_0  index         hin  \
0         0      0       अभी_2   
1         1      1    अभिराम_1   
2         2      2   अभ्यासी_3   
3         3      3   अभ्यासी_3   
4         5      5      आघात_2   
5         6      6      आघात_2   
6         7      7      आघात_2   
7        10     10  अग्रगामी_2   
8        11     11    अहेतुक_1   
9        12     12    अहेतुक_1   
10       13     13     आहुति_1   
11       14     14      अकड़_3   
12       15     15      अकड़_3   
13       16     16      अकड़_3   
14       17     17  आकाशवाणी_1   
15       18     18  आकाशवाणी_1   
16       19     19   आकस्मिक_6   
17       24     26    आलोकित_2   
18       25     27       अमल_5   
19       26     28       अमल_5   

                                             eng_syns hin_id    eng_id  \
0                                   [now, at_present]  11083     49220   
1                                        [attractive]   9106    166146   
2                            [accomplished, complete]     29   2226028   
3   [ace, adept, champion, sensation, maven, mavin...  29929   9762509   
4                                             [wound]  22594   7497122   
5      [abattoir, butchery, shambles, slaughterhouse]   4200   2666943   
6   [beating, thrashing, licking, drubbing, lacing...   8653   1160729   
7                                       [progressive]   2406   1876006   
8         [unproductively, fruitlessly, unprofitably]  17931    214084   
9                             [causeless, reasonless]  16785   1559042   
10                                       [invitation]   5838   7186148   
11  [stubbornness, obstinacy, obstinance, mulishness]  10303   4908835   
12                         [crick, kink, rick, wrick]  13336  14361182   
13                    [ego, egotism, self-importance]   2996   7508806   
14                                    [radio_station]  13989   4044119   
15                                       [revelation]   8506   7213232   
16                                       [unexpected]   2242    930290   
17                                           [bright]   1667    271022   
18  [government, governing, governance, government...   1009   1124794   
19                [faultless, immaculate, impeccable]  11724   1750847   

                                            hin_gloss  \
0                      इसी समय या इस वक़्त या इस क्षण   
1                देखने में अच्छा और मन को लुभाने वाला   
2     जो किसी कार्य को करने में विशेष योग्यता रखता हो   
3   वह जो किसी कार्य को करने की विशेष योग्यता रखता हो   
4                              ठेस का लाक्षणिक प्रयोग   
5                             पशुओं को मारने का स्थान   
6                                     मारने की क्रिया   
7   प्रगति करनेवाला या जो बराबर उन्नति करता हुआ आग...   
8                                        बिना मतलब के   
9                                        बिना कारण का   
10  किसी कार्य में सम्मिलित होने के लिए किसी को आद...   
11   अपनी अनुचित बात पर भी अड़े रहने की अवस्था या भाव   
12                   अकड़ने या ऐंठने की क्रिया या भाव   
13  अपने आपको औरों से बहुत अधिक योग्य, समर्थ या बढ...   
14  वह प्रणाली जिसके तहत विद्युतचुंबकीय तरंगों के ...   
15  वह बात जो ईश्वर की ओर से कही हुई और आकाश से सु...   
16                                  अकस्मात होने वाला   
17               प्रकाश से भरा हुआ या प्रकाश से पूर्ण   
18               राज्य के कार्यों का प्रबंध और संचालन   
19                                जिसमें कोई दोष न हो   

                                            eng_gloss  \
0                             [at the present moment]   
1   [pleasing to the eye or mind especially throug...   
2                                    [highly skilled]   
3    [someone who is dazzlingly skilled in any field]   
4   [a figurative injury (to your feelings or pride)]   
5            [a building where animals are butchered]   
6   [the act of inflicting corporal punishment wit...   
7                    [favoring or promoting progress]   
8                      

In [253]:
fin_df.to_csv('final_linked_set_v3.csv', sep ='\t')
fin_coarse_syn_df.to_csv('final_coarse_linked_set_v1.csv', sep ='\t')